src/nvim/terminal.c

```diff
@@ -250,12 +253,16 @@ Terminal *terminal_open(TerminalOptions opts)
   rv->sb_size = MIN(rv->sb_size, 100000);
   rv->sb_buffer = xmalloc(sizeof(ScrollbackLine *) * rv->sb_size);
 
+  if (!true_color) {
+      return rv;
+  }
+
   // Configure the color palette. Try to get the color from:
   //
   // - b:terminal_color_{NUM}
   // - g:terminal_color_{NUM}
   // - the VTerm instance
```

Comment:
Those two problems won't be solved by this patch if it's what you are asking.

However, it seems like the color_index map might be the cause for those problems. As noted in #2323 there will be some color collisions in the map and the color with the highest index will be returned.

Because it would be best to just return the true index and not just an index that had the same color, perhaps we should remove the color_index map completely.

Instead we will (if the ui is not true color) init all the 256 default colors to (0,0,[0-255]). This should produce an int that is [0-255] so we can the just pass the value to the index value with no lookup needed.